mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp

```diff
@@ -1069,12 +1084,23 @@ class AccessorSubscriptPattern
     return op.getIndex().getType().isa<MemRefType>();
   }
 
-  Value getRef(OpBuilder &builder, Location loc, Type ptrTy, Value acc,
-               Value index) const {
-    const auto ptr = GetMemberPattern<AccessorGetPtr>::loadValue(
-        builder, loc, LLVM::LLVMPointerType::get(ptrTy), acc);
-    return builder.create<LLVM::GEPOp>(loc, ptrTy, ptr, index,
-                                       /*inbounds*/ true);
+  Value getRef(OpBuilder &builder, Location loc, SYCLAccessorSubscriptOp orig,
+               LLVM::LLVMPointerType ptrTy, Value acc, Value index) const {
+    const auto addressSpace =
+        memoryTargetModeToAddressSpace(orig.getAcc()
+                                           .getType()
+                                           .getElementType()
+                                           .cast<AccessorType>()
+                                           .getTargetMode());
+    const auto gepPtrTy =
+        LLVM::LLVMPointerType::get(ptrTy.getElementType(), addressSpace);
+    const auto ptr = GetMemberPattern<AccessorGetPtr>::loadValue(builder, loc,
+                                                                 gepPtrTy, acc);
+    const Value gep = builder.create<LLVM::GEPOp>(loc, gepPtrTy, ptr, index,
+                                                  /*inbounds*/ true);
+    return ptrTy == gepPtrTy
+               ? gep
+               : builder.create<LLVM::AddrSpaceCastOp>(loc, ptrTy, gep);
```

Comment:
Same comment as above about order of `gep` and AS cast.

Reply:
Here, having `gep` before the AS cast matches clang's behavior, therefore this does not need be changed.

Reply:
**Offline conversation**: Changed comment above. Will keep this as is to mimic what `clang` does.